eclipse-solargraph-plugin/src/main/java/io/github/pyvesb/eclipse_solargraph/launch/RubyRunDelegate.java

```diff
@@ -0,0 +1,155 @@
+package io.github.pyvesb.eclipse_solargraph.launch;
+
+import static io.github.pyvesb.eclipse_solargraph.preferences.BooleanPreferences.SYSTEM_RUBY;
+import static io.github.pyvesb.eclipse_solargraph.preferences.StringPreferences.RUBY_DIR;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IAdaptable;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.debug.core.DebugPlugin;
+import org.eclipse.debug.core.ILaunch;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.debug.core.ILaunchConfigurationType;
+import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
+import org.eclipse.debug.core.ILaunchManager;
+import org.eclipse.debug.core.model.LaunchConfigurationDelegate;
+import org.eclipse.debug.ui.ILaunchShortcut;
+import org.eclipse.jface.dialogs.MessageDialog;
+import org.eclipse.jface.viewers.ISelection;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.ui.IEditorInput;
+import org.eclipse.ui.IEditorPart;
+import org.eclipse.ui.PlatformUI;
+
+public class RubyRunDelegate extends LaunchConfigurationDelegate implements ILaunchShortcut {
+
+	public static final String PROGRAM = "PROGRAM";
+	public static final String ARGUMENTS = "ARGUMENTS";
+
+	@Override
+	public void launch(ILaunchConfiguration configuration, String mode, ILaunch launch, IProgressMonitor monitor)
+			throws CoreException {
+ 		String fileName = configuration.getAttribute(PROGRAM, "");
+		String argument = configuration.getAttribute(ARGUMENTS, "");
+		String workingDir = configuration.getAttribute(DebugPlugin.ATTR_WORKING_DIRECTORY, "");
+		DebugPlugin.getDefault().getLaunchManager().addLaunch(launch);
+		String[] command = getFileCommand(fileName, argument);
+		String commandSummary = "ruby " + fileName;
+		Job.create("Running " + commandSummary, r -> {
+			Process process = DebugPlugin.exec(command, new File(workingDir));
+			DebugPlugin.newProcess(launch, process, commandSummary);
+		}).schedule();
+	}
+
+	private String[] getFileCommand(String fileName, String arguments) {
+		StringBuilder ruby = new StringBuilder();
+		if (!SYSTEM_RUBY.getValue()) {
+			ruby.append(RUBY_DIR.getValue());
+			if (ruby.length() > 0) {
+				ruby.append(File.separator);
+			}
+		}
+		ruby.append("ruby");
+		List<String> command = new ArrayList<>();
+		command.add(ruby.toString());
+		command.add(fileName);
+		command.addAll(Arrays.asList(arguments.split(" ")));
+		return command.toArray(new String[0]);
+	}
+
+	@Override
+	public void launch(ISelection selection, String mode) {
+		if (selection instanceof IStructuredSelection) {
```

Comment:
When is this condition false? _plugin.xml_ currently enforces a selection of one element of type `IFile`, when is that not `IStructuredSelection`?

Reply:
This is just safety net. We have no idea when/whether some view uses the Run As popup menu and how it populates the selection. E.g. Search view shows Run As and has "Show As List" option - in this case it still returns IStructuredSelection but it could have been ISelection, there is no guarantee that some other view doesn't do it though. I can remove it if you want but this kind of coding prevents surprises in the endless combinations of views users ends up with.